flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetReaders.java

```diff
@@ -19,64 +19,678 @@
+      if (readPos < map.capacity()) {
+        entry.set((K) map.keys.values[readPos], (V) map.values.values[readPos]);
+        kv = entry;
+      }
+
+      readPos += 1;
+
+      return kv;
+    }
+
+    @Override
+    protected void addPair(ReusableMapData map, K key, V value) {
+      if (writePos >= map.capacity()) {
+        map.grow();
+      }
+
+      map.keys.values[writePos] = key;
+      map.values.values[writePos] = value;
+
+      writePos += 1;
+    }
+
+    @Override
+    protected MapData buildMap(ReusableMapData map) {
+      map.setNumElements(writePos);
+      return map;
+    }
   }
 
-  private static class RowReader extends ParquetValueReaders.StructReader<Row, Row> {
-    private final Types.StructType structType;
+  private static class RowDataReader extends ParquetValueReaders.StructReader<RowData, GenericRowData> {
+    private final int numFields;
 
-    RowReader(List<Type> types, List<ParquetValueReader<?>> readers, Types.StructType struct) {
+    RowDataReader(List<Type> types, List<ParquetValueReader<?>> readers) {
       super(types, readers);
-      this.structType = struct;
+      this.numFields = readers.size();
     }
 
     @Override
-    protected Row newStructData(Row reuse) {
-      if (reuse != null) {
-        return reuse;
+    protected GenericRowData newStructData(RowData reuse) {
+      if (reuse instanceof GenericRowData) {
+        return (GenericRowData) reuse;
       } else {
-        return new Row(structType.fields().size());
+        return new GenericRowData(numFields);
       }
     }
 
     @Override
-    protected Object getField(Row row, int pos) {
-      return row.getField(pos);
+    protected Object getField(GenericRowData intermediate, int pos) {
+      return intermediate.getField(pos);
+    }
+
+    @Override
+    protected RowData buildStruct(GenericRowData struct) {
+      return struct;
+    }
+
+    @Override
+    protected void set(GenericRowData row, int pos, Object value) {
+      row.setField(pos, value);
+    }
+
+    @Override
+    protected void setNull(GenericRowData row, int pos) {
+      row.setField(pos, null);
+    }
+
+    @Override
+    protected void setBoolean(GenericRowData row, int pos, boolean value) {
+      row.setField(pos, value);
+    }
+
+    @Override
+    protected void setInteger(GenericRowData row, int pos, int value) {
+      row.setField(pos, value);
+    }
+
+    @Override
+    protected void setLong(GenericRowData row, int pos, long value) {
+      row.setField(pos, value);
     }
 
     @Override
-    protected Row buildStruct(Row row) {
-      return row;
+    protected void setFloat(GenericRowData row, int pos, float value) {
+      row.setField(pos, value);
     }
 
     @Override
-    protected void set(Row row, int pos, Object value) {
+    protected void setDouble(GenericRowData row, int pos, double value) {
       row.setField(pos, value);
     }
   }
+
+  private static class ReusableMapData implements MapData {
```

Comment:
I think we can use the same ReusableMapData class for both Spark and Flink, since they are so similar. Some of the data access methods in the concrete classes would need to be implemented in a concrete class the engine-specific interface, but the main parts could be shared. And I think that would mean we could share significant parts of the reader classes as well.

I don't think we need to do this refactor now, but we should try to deduplicate this.

Reply:
Opened https://github.com/apache/iceberg/issues/1331.